In [0]:
import pandas as pd
import pandas
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
from sklearn.linear_model import LassoLarsCV
import matplotlib.pyplot as plt
from google.colab import drive 
drive.mount('/content/gdrive')

Drive already mounted at /content/gdrive; to attempt to forcibly remount, call drive.mount("/content/gdrive", force_remount=True).


In [0]:
df=pd.read_csv('gdrive/My Drive/finalmaster-ratios.csv')
df.head()

# Purchases  B01001001   B01001002  ...  B19001015  B19001016  B19001017
0           22     206252  469.226965  ...  36.440765  23.446284  21.197485
1            7      61399  486.538869  ...  33.000508  33.169741  24.792689
2            3      73170  489.859232  ...  20.125056  11.890525  16.537397
3           94     251724  505.585483  ...  43.731067  38.851729  40.427349
4            0      37382  495.586111  ...  36.118530  31.603714  19.648989

[5 rows x 190 columns]

In [0]:
# create a list of all the predictors you're going to feed into the LassoLarsCV model
allvariablenames = list(df.columns.values)
listofallpredictors = allvariablenames[8:]

In [0]:
#load predictors into dataframe
predictors = df[listofallpredictors]  
#load target into dataframe
target = df['# Purchases']   
target.head()

0    22
1     7
2     3
3    94
4     0
Name: # Purchases, dtype: int64

In [0]:
# split data into train and test sets, with 30% retained for test
pred_train, pred_test, tar_train, tar_test = train_test_split(predictors, target, test_size=.3, random_state=123) 
#  (1) pred_train, the predictors training set 
#  (2) pred_test, the predictors test test 
#  (3) tar_train, the target training set 
#  (4) tar_test, the target test set.
model = LassoLarsCV(cv=10, precompute=False).fit(pred_train, tar_train)

/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/least_angle.py:578: ConvergenceWarning: Regressors in active set degenerate. Dropping a regressor, after 5 iterations, i.e. alpha=1.496e+00, with an active set of 5 regressors, and the smallest cholesky pivot element being 6.322e-08. Reduce max_iter or increase eps parameters.
  ConvergenceWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/least_angle.py:578: ConvergenceWarning: Regressors in active set degenerate. Dropping a regressor, after 8 iterations, i.e. alpha=1.098e+00, with an active set of 8 regressors, and the smallest cholesky pivot element being 6.322e-08. Reduce max_iter or increase eps parameters.
  ConvergenceWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/least_angle.py:578: ConvergenceWarning: Regressors in active set degenerate. Dropping a regressor, after 10 iterations, i.e. alpha=7.329e-01, with an active set of 10 regressors, and the smallest cholesky pivot element b

## Question 1:In your own words, explain what the above lines of code are doing.Why am I doing it? Explain each line.

In [0]:
# build coefficent chart
# Create a predictors model data frame named 'predictors' that contains the list of all possible predictor variables (load all predictors to this model)
predictors_model=pd.DataFrame(listofallpredictors)
# named the column labels of predictors_model to 'label'
predictors_model.columns = ['label']
# add a new column called 'coeff' and append all coefficents from regression model.
predictors_model['coeff'] = model.coef_
# check all the index and row in predictors
for index, row in predictors_model.iterrows():
    # for loop that go through predictor_model table and print out the coefficent with name that larger than zero.
    if row['coeff'] > 0:
        # print out the significant values
        print(row.values)

['B01001014' 0.8558761066941788]
['B01001036' 2.5053482381631653]
['B01001037' 0.8892493223320962]
['B01001038' 1.5316387928880384]
['B02001005' 0.41252295298457853]
['B13014026' 0.48004105312075906]
['B13014027' 0.6978957445987839]
['B13016001' 875149895.329212]
['B19001017' 1.4834348068681533]


## Question 2: Interpret each variable intuitively. What Census variables most predict sales? What does that mean, practically? Here's what I'd put for the example above. "In areas where there are more females aged 30-34, we sell more Bobo Bars."


['B01001014' 0.8557908775529921] Males aged 40 to 44 Years.
In areas where there are more males aged 40-44, we sell 0.8557908775529921 unit more Bobo Bars

['B01001036' 2.505392496591849] Females aged 30 to 34 Years.
In areas where there are more females aged 30-34, we sell 2.505392496591849 unit more Bobo Bars

['B01001037' 0.8894214357013622] Females aged 35 to 39 Years.
In areas where there are more females aged 35-39, we sell 0.8894214357013622] unit more Bobo Bars

['B01001038' 1.5315839680821497] Females aged 40 to 44 Years.
In areas where there are more females aged 40-44, we sell 1.5315839680821497 unit more Bobo Bars

['B02001005' 0.4125408937426837] Asian Alone.
In areas where there are asian alone, we sell 0.4125408937426837 unit more Bobo Bars

['B13014026' 0.4800240326923769] Women 15 to 50 Years Who Had a Birth in the Past 12 Months with Bachelor's Degree.
In areas where there are more females aged 15-50 that had a birth in the past 12 months with Bachelor's degree, we sell 0.4800240326923769 unit more Bobo Bars

['B13014027' 0.6977454940063235] Women 15 to 50 Years Who Had a Birth in the Past 12 Months with Graduate or Professional Degree.
In areas where there are more females aged 15-50 that had a birth in the past 12 months with graduate or professional degree, we sell 0.6977454940063235 unit more Bobo Bars

['B13016001' 874922971.7249781] Women 15 to 50 Years Who Had a Birth in the Past 12 Months.
In areas where there are more females aged 15-50 that had a birth in the past 12 months, we sell 874922971.7249781 unit more Bobo Bars

['B19001017' 1.4834465563617387] Household with income 200,000 or More.
In areas where there are more household with income 200,000 or more, we sell 1.4834465563617387 unit more Bobo Bars

## Question 3: If I had to report only two census variables to my boss that most steeply predicted sales, what would those be?

Since the higher the value of the coefficient, the steeper the relationship between sales and that variable.
I will report B01001036 which is Females aged 30 to 34 Years, 
and  B13016001 which is Women 15 to 50 Years Who Had a Birth in the Past 12 Months because these 2 census have the most 2 highest value of the coefficient which means the most steep relation between sales and attributes.

## Question 4:Are the training and text set mean squared errors similar? What does that mean practically?

In [0]:
train_error = mean_squared_error(tar_train, model.predict(pred_train))
print ('training data MSE')
print(train_error)

training data MSE
22025.491066757


In [0]:
train_error_test = mean_squared_error(tar_test, model.predict(pred_test))
print ('testing data MSE')
print(train_error_test)

testing data MSE
41549.54803776253


The mean squared error of training data set is 22025.491066757
and the mean squared error of testing data is 41549.54803776253 which are totaly different. The MEAN SQUARED ERROR is a measure of how close a fitted line is to data points,which means smaller mean squared error represents the closer fit to the data or model. 
From our example, the training data has smaller mean squard error which means fitting batter for our model because our model based on training data, so we also can say the model we builded might not fit testing data because MSE of testing  data is much larger which is make sense. 

### Then we do R squre

In [0]:
#r squared
rsquared_train=model.score(pred_train,tar_train)
print ('training data R-square')
print(rsquared_train)

training data R-square
0.2400221219784492


In [0]:
#r squared
rsquared_test=model.score(pred_test,tar_test)
print ('testing data R-square')
print(rsquared_test)

testing data R-square
0.1758628512005107


R-squared is a statistical measure that represents the proportion of the variance for a dependent variable that's explained by an independent variable or variables in a regression model. Whereas correlation explains the strength of the relationship between an independent and dependent variable, R-squared explains to what extent the variance of one variable explains the variance of the second variable.
So lagerer R square represents the closer fit to the data or model. From our example,The R-square of traninig data is 0.2400221219784492 and R-square of testing data is 0.1758628512005107, The training data is a better regression model as it has larger R-square value according to we build modle based on trainong data which is make sense.



## Question 5: If your boss asked, "How well does Census data, overall, predict sales?" What would you say? Why?

I would say census is not a good data for predicting sales.
From our processing for meaning squared error and R-squard, we found both of training data are closer fitting to the model because we built model cased on training data. The problem is we got a pretty small R-square value and a large meaning squard error value for both data (training data and testing data) which means not sufficient enough to make proper prediction. We can find the sd of the regression model’s errors is only about 1/10 the size of the sd of the errors that you would get with a constant-only model. Because we only got R- square value around 20% which means the variance of its errors is 20% less than the variance of the dependent variable and the sd of its error is 11% less than the sd of the dependent variable.At the same time, the value of MSE are pretty large, so its not good.


## Question 6: What is our baseline sales number? What does that mean, practically? 

In [0]:
print("y interecept:")
print(model.intercept_)

y interecept:
22.19738813257551


 In this case, we could know that the baseline sales number is 22.194675187960804,because when X=0, the corresponding y-value is the y-intercept which is 22.19738813257551. which means when variables that truly predict unique amounts of sales equal to 0, the baseline sales are 22.194675187960804.